src/subcommands/deployments.ts

```diff
@@ -334,6 +376,183 @@ async function deleteDeployment(args: Args): Promise<void> {
+    mProject === undefined && mProjectId
+      ? api.getProject(mProjectId)
+      : undefined,
+    mProjectId ? api.getProjectDatabases(mProjectId) : undefined,
+  ]);
+  if (!mBuild) {
+    spinnerPrep.fail(
+      `The deployment '${deploymentId}' does not exist, or you don't have access to it`,
+    );
+    return Deno.exit(1);
+  }
+  const build = mBuild;
+  const projectId = build.project.id;
+  if (mProject === undefined || mDatabases === undefined) {
+    // We didn't have projectId before. Now we do
+    [mProject, mDatabases] = await Promise.all([
+      mProject ? Promise.resolve(mProject) : api.getProject(projectId),
+      mDatabases
+        ? Promise.resolve(mDatabases)
+        : api.getProjectDatabases(projectId),
+    ]);
+  }
+  if (!mProject) {
+    spinnerPrep.fail(
+      `The project '${projectId}' does not exist, or you don't have access to it`,
+    );
+    return Deno.exit(1);
+  }
+  const project = mProject;
+  const databases = mDatabases;
+
+  // ----
+
+  const alreadyProd =
+    project.productionDeployment?.deploymentId === build.deploymentId;
+  const prod = args.prod ?? alreadyProd;
+  const prodDatabase = databases?.find((database) =>
+    deploymentDatabaseEnv(project, database) === "Production"
+  );
+  const previewDatabase = databases?.find((database) =>
+    deploymentDatabaseEnv(project, database) === "Preview"
+  );
+  let db;
+  switch (args.db?.toLowerCase().trim()) {
+    case "prod":
+    case "production": {
+      if (!prodDatabase) {
+        spinnerPrep.fail(
+          `Project '${project.name}' does not have a production database`,
+        );
+        return Deno.exit(1);
+      }
+      db = prodDatabase.databaseId;
+      break;
+    }
+    case "preview": {
+      if (!previewDatabase) {
+        spinnerPrep.fail(
+          `Project '${project.name}' does not have a preview database`,
+        );
+        return Deno.exit(1);
+      }
+      db = previewDatabase.databaseId;
+      break;
+    }
+    default:
+      db = args.db;
+  }
+
+  if (!db) {
+    // For GitHub deployments, Deploy assigns the branch database also during redeployment
+    // Unless the user is explicit about the db, we want to maintain the invariant status == databaseEnv
+    if (prod) {
+      db = prodDatabase?.databaseId;
+    } else {
+      db = previewDatabase?.databaseId;
+    }
+  }
+
+  const envVarsToAdd = await envVarsFromArgs(args) || {};
+  const addedEnvs = Object.keys(envVarsToAdd);
+  // If the redeployment sets some env vars, the remaining env vars in the deployment are deleted
+  const envVarsToRemove = build.deployment && addedEnvs.length > 0
+    ? Object.fromEntries(
+      build.deployment.envVars
+        .filter((env) => !addedEnvs.includes(env))
+        // HOME is always set by Deno Deploy
+        .filter((env) => env !== "HOME")
+        .map((key) => [key, null]),
+    )
+    : {};
+  const removedEnvs = Object.keys(envVarsToRemove);
```

Comment:
Question: is there any particular reason that we don't use the default logic of merging the passed env vars with those of the base deployment?

Reply:
I've been going back and forth with this. Ultimately I settled with this approach because:
1. Otherwise, there wouldn't be any mechanism to remove envs from deployments
2. Right now It's impossible to know which env var values the base deployment uses, thus if we merge the base deployment's envs with the provided envs, it's impossible to fully know which env var set a deployment will end up with. This approach gives back full control to the user. 
3. We want to promote the management of env vars entirely from the local environment / GH secrets (for GH action deployments), and see if this becomes a better DX than setting project env variables in Deno Deploy, which so far has been a significant source of friction. Project env variables are really only valuable when doing automatic GH deployments. If envs are managed locally already, it's not a problem to provide the full env var set on each [re]deployment, and It's a simpler and less confusing experience.

Reply:
Thanks for the explanation. Makes kinda sense to me, especially I even missed the first point (dotenv files are not expressive enough to represent `null` that we use for deleting an existing env var)